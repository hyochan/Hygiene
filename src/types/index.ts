import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';
import firebase from 'firebase/app';

export type FirebaseUser = firebase.User;

export enum SocialAuthProvider {
  Google = 'google',
  Facebook = 'facebook',
  Apple = 'apple',
};

export type User = {
  uid: string;
  displayName: string | null;
  email: string;
  introduction?: string | null;
  point: number;
  photoURL: string | null;
  thumbURL?: string;
  signInMethod?: string | null;
}

export type Reply = {
  id: string;
  text: string;
  writerId: string;
  createdAt?: Date;
  updatedAt?: Date;
}

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export interface Activity {
  id: string;
  writerId: string;
  message: string;
  urls: string[];
  category: string;
  createdAt?: Date;
  updatedAt?: Date;
}

export interface Photo {
  uri: string;
  thumbUri?: string;
  added?: boolean;
  deleted?: boolean;
}

export enum CategoryType {
  HandWash = 'HandWash',
  WearMask = 'WearMask',
  StayHome = 'StayHome',
  GoodConsumption = 'GoodConsumption',
  Etc = 'Etc',
}

export interface Category {
  type: string;
  point: number;
  text: string;
  svgIcon: any;
  color: string;
}

export enum SelectImageActionType {
  LAUNCH_CAMERA = 0,
  LAUNCH_GALLERY = 1,
  CANCEL = 2,
}

export enum SelectMyFeedMoreActionType {
  DELETE = 0,
  CANCEL = 1,
}

export enum SelectPeerFeedMoreActionType {
  CANCEL = 0,
}

export interface Throphy {
  type: string;
  text: string;
  count: number;
}
