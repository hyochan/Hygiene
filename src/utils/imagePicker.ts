import * as ImageManipulator from 'expo-image-manipulator';
import * as ImagePicker from 'expo-image-picker';
import * as Permissions from 'expo-permissions';

import { Image, Platform } from 'react-native';

export const THUMBNAIL_SIZES = {
  MAX_WIDTH: 192,
  MAX_HEIGHT: 192,
};

export interface ImageSize {
  width: number;
  height: number;
}

export const getOriginalImageSize = async (imageUri: string): Promise<ImageSize> => new Promise<ImageSize>(
  (resolve, reject) => Image.getSize(
    imageUri,
    (width: number, height: number) => resolve({
      width,
      height,
    }),
    reject,
  ),
);

export const resizeImage = async ({
  imageUri,
  maxWidth,
  maxHeight,
}: {
  imageUri: string;
  maxWidth?: number;
  maxHeight?: number;
}): Promise<ImageManipulator.ImageResult> => {
  const manipulate = ({
    width,
    height,
  }: {
    width?: number;
    height?: number;
  }): Promise<ImageManipulator.ImageResult> => ImageManipulator.manipulateAsync(
    imageUri,
    [{ resize: { width, height } }],
    { compress: 1, format: ImageManipulator.SaveFormat.PNG },
  );

  const { width: originWidth, height: originHeight } = await getOriginalImageSize(imageUri);

  if (originWidth >= originHeight) {
    if (maxWidth === undefined || maxWidth === null) {
      if (originWidth < THUMBNAIL_SIZES.MAX_WIDTH) {
        return manipulate({ width: originWidth });
      }

      return manipulate({ width: THUMBNAIL_SIZES.MAX_WIDTH });
    }

    if (originWidth < maxWidth) {
      return manipulate({ width: originWidth });
    }

    return manipulate({ width: maxWidth });
  }

  if (maxHeight === undefined || maxHeight === null) {
    if (originHeight < THUMBNAIL_SIZES.MAX_HEIGHT) {
      return manipulate({ height: originHeight });
    }

    return manipulate({ height: THUMBNAIL_SIZES.MAX_HEIGHT });
  }

  if (originHeight < maxHeight) {
    return manipulate({ height: originHeight });
  }

  return manipulate({ height: maxHeight });
};

enum MediaTypeOptions {
  All = 'All',
  Videos = 'Videos',
  Images = 'Images',
}

enum PermissionStatus {
  UNDETERMINED = 'undetermined',
  GRANTED = 'granted',
  DENIED = 'denied',
}

const photoOptions = {
  mediaTypes: ImagePicker.MediaTypeOptions.Images,
  allowsEditing: true,
  exif: true,
};

const requestPermissions = async (
  type: string,
): Promise<Permissions.PermissionStatus> => {
  if (type === 'photo') {
    const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);

    return status;
  }

  const { status } = await Permissions.askAsync(Permissions.CAMERA);

  return status;
};

export const launchCameraAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  const permissionStatus = await requestPermissions('camera');

  if (permissionStatus === PermissionStatus.GRANTED) {
    return ImagePicker.launchCameraAsync(photoOptions);
  }

  return null;
};

export const launchImageLibraryAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  if (Platform.OS === 'web') {
    return ImagePicker.launchImageLibraryAsync(photoOptions);
  }

  const permissionStatus = await requestPermissions('photo');

  if (permissionStatus === PermissionStatus.GRANTED) {
    return ImagePicker.launchImageLibraryAsync(photoOptions);
  }

  return null;
};
