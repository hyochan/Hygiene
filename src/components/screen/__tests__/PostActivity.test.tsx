import 'react-native';

import React, { ReactElement } from 'react';
import { RenderAPI, render } from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import { CategoryType } from '../../../types';
import Screen from '../PostActivity';
import { categoryMap } from '../../../utils/constants';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: ReactElement;
let testingLib: RenderAPI;

describe('Rendering', () => {
  beforeEach(() => {
    props = createTestProps({
      route: {
        params: {
          category: categoryMap[CategoryType.HandWash],
        },
      },
    });

    component = createTestElement(<Screen {...props} />);
    testingLib = render(component);
  });

  it('renders without crashing', () => {
    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});

describe('Interaction', () => {
  beforeEach(() => {
    props = createTestProps({
      route: {
        params: {
          category: categoryMap[CategoryType.HandWash],
        },
      },
    });

    component = createTestElement(<Screen {...props} />);
    testingLib = render(component);
  });

  it('should simulate onClick', () => {
    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    // const btn = testingLib.queryByTestId('btn');
    // act(() => {
    //   fireEvent.press(btn);
    //   fireEvent.press(btn);
    // });
    // expect(cnt).toBe(3);
  });
});
