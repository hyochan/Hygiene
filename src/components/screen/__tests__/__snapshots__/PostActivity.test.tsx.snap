// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interaction should simulate onClick 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#ffffff",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginTop": 16,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginBottom": 8,
          }
        }
      >
        Washing hands
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flexDirection": "column",
          },
          Object {
            "height": 187,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#b9b9c4",
              "fontSize": 14,
              "fontWeight": "500",
              "marginBottom": 5,
            },
            undefined,
            null,
          ]
        }
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        multiline={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Please leave what you want to say."
        placeholderTextColor="#808080"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "fontSize": 15,
              "fontWeight": "500",
              "paddingBottom": 15,
              "paddingTop": 15,
            },
            Object {
              "backgroundColor": "#f7f7f7",
              "color": "black",
              "fontSize": 14,
              "height": 187,
              "padding": 12,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E6E6E6",
            "height": 1,
            "marginBottom": 10,
            "marginTop": 20,
            "width": "100%",
          },
        ]
      }
    />
    <View
      style={
        Object {
          "marginTop": 12,
          "paddingHorizontal": 22,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "flex-start",
              "paddingBottom": 0,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 0,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderColor": "#E6E6E6",
                "borderRadius": 10,
                "borderWidth": 1,
                "height": 80,
                "justifyContent": "center",
                "width": 80,
              }
            }
          >
            <SvgMock
              stroke="#444444"
            />
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 14,
              "marginTop": 18,
            },
          ]
        }
      >
        Inappropriate content, such as irrelevant activities or repetition of the same character, may be deleted without prior warning.
      </Text>
    </View>
  </View>
</View>
`;

exports[`Rendering renders without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#ffffff",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginTop": 16,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
            "marginBottom": 8,
          }
        }
      >
        Washing hands
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flexDirection": "column",
          },
          Object {
            "height": 187,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#b9b9c4",
              "fontSize": 14,
              "fontWeight": "500",
              "marginBottom": 5,
            },
            undefined,
            null,
          ]
        }
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        multiline={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Please leave what you want to say."
        placeholderTextColor="#808080"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "fontSize": 15,
              "fontWeight": "500",
              "paddingBottom": 15,
              "paddingTop": 15,
            },
            Object {
              "backgroundColor": "#f7f7f7",
              "color": "black",
              "fontSize": 14,
              "height": 187,
              "padding": 12,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E6E6E6",
            "height": 1,
            "marginBottom": 10,
            "marginTop": 20,
            "width": "100%",
          },
        ]
      }
    />
    <View
      style={
        Object {
          "marginTop": 12,
          "paddingHorizontal": 22,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "flex-start",
              "paddingBottom": 0,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 0,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderColor": "#E6E6E6",
                "borderRadius": 10,
                "borderWidth": 1,
                "height": 80,
                "justifyContent": "center",
                "width": 80,
              }
            }
          >
            <SvgMock
              stroke="#444444"
            />
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 14,
              "marginTop": 18,
            },
          ]
        }
      >
        Inappropriate content, such as irrelevant activities or repetition of the same character, may be deleted without prior warning.
      </Text>
    </View>
  </View>
</View>
`;
